File: practice/algorithms/implementation/sock_merchant/SockMerchant.js
'use strict';

const processData = input => {
    const lines = input.split('\n');
    const socks = lines[1].split(' ').map(i => parseInt(i));
    const sockGroupsCount = new Map();
    for(let i = 0; i < socks.length; i++) {
        const color = socks[i];
        if(sockGroupsCount.get(color)) {
            const count = sockGroupsCount.get(color);
            sockGroupsCount.set(color, count + 1);
        } else {
            sockGroupsCount.set(color, 1);
        }
    }
    let pairs = 0;
    const groups = Array.from(sockGroupsCount.values());
    for(let i = 0; i < groups.length; i++) {
        const count = groups[i];
        pairs += Math.floor(count / 2);
    }

    console.log(pairs);
};

process.stdin.resume();
process.stdin.setEncoding("ascii");

let _input = "";
process.stdin.on("data", input => _input += input);
process.stdin.on("end", () => processData(_input));

File: practice/algorithms/implementation/sock_merchant/SockMerchant.java
import java.util.*;

public class SockMerchant {

    public static void main(String[] args) {
        Scanner stdin = new Scanner(System.in);
        int sockCount = stdin.nextInt();
        Map<Integer, Integer> sockGroupsCount = new HashMap<>();
        for(int i = 0; i < sockCount; i++) {
            int color = stdin.nextInt();
            if(sockGroupsCount.containsKey(color)) {
                int count = sockGroupsCount.get(color);
                sockGroupsCount.put(color, count + 1);
            } else {
                sockGroupsCount.put(color, 1);
            }
        }
        int pairs = 0;
        for(int count : sockGroupsCount.values()) {
            pairs += count / 2;
        }

        System.out.println(pairs);
        stdin.close();
    }
}